server/src/main/kotlin/fi/oph/kitu/auth/WebSecurityConfig.kt

```diff
@@ -29,44 +30,37 @@ class WebSecurityConfig {
         casConfig: CasConfig,
         environment: Environment,
     ): SecurityFilterChain {
-        http
-            .csrf { csrf ->
-                csrf.ignoringRequestMatchers("/api/*", "/db-scheduler-api/**")
-            }.logout {
-                it.logoutSuccessUrl(casConfig.getCasLogoutUrl())
-                it.logoutUrl("/logout")
-                it.logoutRequestMatcher(AntPathRequestMatcher("/logout", "GET"))
-            }.authorizeHttpRequests { authorize ->
-                authorize
-                    // Allow healthcheck without login
-                    .requestMatchers(
-                        "/actuator/health",
-                    ).permitAll()
-                    // Make Swagger UI and OpenAPI schema YAML publicly available
-                    .requestMatchers(
-                        "/swagger-ui.html",
-                        "/swagger-ui/*",
-                        "/v3/api-docs/swagger-config",
-                        "/open-api.yaml",
-                    ).permitAll()
-            }.addFilter(casAuthenticationFilter)
```

Comment:
Ilmeisesti tuo metodi automaagisesti järjestää filtterit tämän järjestyksen perusteella: https://github.com/spring-projects/spring-security/blob/e947678424c939e924fdcb2205534dec370617ac/config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterOrderRegistration.java#L76-L127

Reply:
...ja tässä tiedostossa joka määrittelee Kotlin-DSL:n ei tarjota tuota `addFilter()`-metodia ollenkaan: https://github.com/spring-projects/spring-security/blob/main/config/src/main/kotlin/org/springframework/security/config/annotation/web/HttpSecurityDsl.kt

Reply:
Vedän taas copy pastalla ->

Reply:
Ehkä toimii nyt.

Reply:
You're correct that the Kotlin DSL for Spring Security does not directly expose `addFilter`, `addFilterBefore`, or `addFilterAfter` as methods. This is because the Kotlin DSL is designed to simplify common configurations and doesn't include every feature of the `HttpSecurity` API.

To add a custom filter when using the Kotlin DSL, you can fall back to using the traditional `HttpSecurity` methods within the DSL block. Here's how you can integrate your custom filter:

---

### **Revised Example with Custom Filter**
You can use `http.addFilter` (or its alternatives) inside the Kotlin DSL block by accessing the traditional `HttpSecurity` API directly:

```kotlin
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter

@Configuration
@EnableWebSecurity
class PartialSecurityConfig {

    @Bean
    open fun userDetailsService(): UserDetailsService {
        // Provide a UserDetailsService implementation
    }

    @Bean
    open fun authenticationManager(http: HttpSecurity): AuthenticationManager {
        return http.getSharedObject(AuthenticationManager::class.java)
    }

    @Bean
    open fun casAuthenticationFilter(authManager: AuthenticationManager): CasAuthenticationFilter {
        val filter = CasAuthenticationFilter()
        filter.setAuthenticationManager(authManager)
        filter.setFilterProcessesUrl("/cas/login") // Example filter URL
        return filter
    }

    @Bean
    open fun securedFilterChain(http: HttpSecurity): SecurityFilterChain {
        http {
            securityMatcher("/secured/**")
            authorizeHttpRequests {
                authorize("/secured/user", hasRole("USER"))
                authorize("/secured/admin", hasRole("ADMIN"))
                authorize(anyRequest, authenticated)
            }
            httpBasic { }
            formLogin { }

            // Use traditional HttpSecurity methods to add the custom filter
            http.addFilterBefore(casAuthenticationFilter(authenticationManager(http)), UsernamePasswordAuthenticationFilter::class.java)
        }
        return http.build()
    }
}
```

---

### **Explanation of Changes**

1. **Fallback to Traditional API**:
   - The Kotlin DSL (`http { ... }`) simplifies common configurations but doesn't support every `HttpSecurity` method.
   - To add a filter, you can directly call `http.addFilterBefore`, `http.addFilterAfter`, or `http.addFilter` inside the DSL block.

2. **Specify the Filter Position**:
   - Use `addFilterBefore` to place your filter before a specific filter, such as `UsernamePasswordAuthenticationFilter`.
   - Alternatives:
     - `addFilterAfter` places it after a specified filter.
     - `addFilter` appends it to the end of the chain.

3. **Custom Filter Bean**:
   - The `casAuthenticationFilter` is defined as a Spring bean to ensure it can be reused and configured properly (e.g., setting an `AuthenticationManager`).

4. **Authentication Manager Dependency**:
   - Ensure the `AuthenticationManager` is shared with the custom filter to handle authentication.

---

### **Why Is There No `addFilter` in the Kotlin DSL?**
The Kotlin DSL is a convenience layer that covers the most common configurations. It doesn't attempt to wrap every possible method of `HttpSecurity`. For advanced use cases like adding filters, falling back to traditional Java-style API calls is the recommended approach.

---

### **Final Notes**
- This approach ensures you can seamlessly add custom filters while leveraging the simplicity of the Kotlin DSL for other configurations.
- If you need further customization or clarification, let me know!

Reply:
Hieman väärin ChatGPT tuossakin sanoi, sillä `addFilterBefore` ja `-After` on kyllä DSL:ssä, vain `addFilter` puuttuu.